enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/AuthFilter.java

```diff
@@ -134,17 +147,36 @@ public boolean handleRequest(RequestContext requestContext) {
             return true;
         }
 
+        boolean authenticated = false;
         boolean canAuthenticated = false;
         for (Authenticator authenticator : authenticators) {
             if (authenticator.canAuthenticate(requestContext)) {
-                canAuthenticated = true;
+                if (!authenticator.getName().contains(APIConstants.API_SECURITY_MUTUAL_SSL_NAME)) {
+                    canAuthenticated = true;
+                }
                 AuthenticationResponse authenticateResponse = authenticate(authenticator, requestContext);
-                if (authenticateResponse.isAuthenticated() && !authenticateResponse.isContinueToNextAuthenticator()) {
+                if (authenticateResponse.isMandatoryAuthentication()) {
+                    authenticated = authenticateResponse.isAuthenticated();
```

Comment:
No. Let's say, application-level security is mandatory. Then only one of the application-level authenticators is needed to be authenticated. Therefore, we need to check authenticators until we find an authenticated one. If the first authenticator is unauthenticated, we cannot return an error since the next one may be authenticated. This logic says, to update 'authenticated' only if the authentication is a mandatory one. [carbon-apimgt code](https://github.com/wso2/carbon-apimgt/blob/604205fb734b449603d63f519eaa6137993a4586/components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/APIAuthenticationHandler.java#L546)